### PR TITLE
Updating `optimize` to handle multiple risk constraints

### PR DESCRIPTION
**Pull request by anirban-chaudhuri**
- Adding support for multiple risk constraints: requires `risk_bound`, `alpha`, and `qoi` to be List.

### CODE REVIEW

File: pyciemss/ouu/ouu.py [hunk 1]
```diff
@@ -66,19 +66,19 @@ def __init__(
         self,
         model: Callable,
         interventions: Callable[[torch.Tensor], Dict[float, Dict[str, Intervention]]],
-        qoi: Callable,
+        qoi: List[Callable[[Any], np.ndarray]],
         end_time: float,
         logging_step_size: float,
         *,
         start_time: float = 0.0,
-        risk_measure: Callable = lambda z: alpha_superquantile(z, alpha=0.95),
```
Comment (SamWitty) [inline pyciemss/ouu/ouu.py:74]: Nitpick: Could we rename this argument `risk_measures`?

File: pyciemss/ouu/ouu.py [hunk 2]
```diff
@@ -107,15 +107,17 @@ def __call__(self, x):
...
             # Apply intervention and perform forward uncertainty propagation
             samples = self.propagate_uncertainty(x)
             # Compute quanity of interest
-            sample_qoi = self.qoi(samples)
+            sample_qoi = [q(samples) for q in self.qoi]
             # Estimate risk
-            risk_estimate = self.risk_measure(sample_qoi)
+            risk_estimate = np.full(len(self.qoi), np.nan)
+            for i in range(len(self.qoi)):
+                risk_estimate[i] = self.risk_measure[i](sample_qoi[i])
```
Comment (SamWitty) [inline pyciemss/ouu/ouu.py:120]: @anirban-chaudhuri , how is this now expanded risk_estimate array used downstream? I'm having trouble tracking how this propagates to the optimizer, and what the implication is for these multiple risk measures?

My understanding based on our many conversations and earlier reviews is that these risk measures are only used for constraints, and that the scipy optimizer accepts arrays as constraints, checking if **any** element is violated. Is that correct?

File: docs/source/interfaces.ipynb
Comment (SamWitty) [inline docs/source/interfaces.ipynb:1]: This looks good. My only request is to remove `squeeze` where possible, and to remove all commented out code.

See the more thorough discussion in the main body of this review for discussion of the `squeeze` issue.

File: docs/source/optimize_interface.ipynb
Comment (SamWitty) [inline docs/source/optimize_interface.ipynb:1]: See comment from `docs/source/interfaces.ipynb` about `squeeze`.

**Review by SamWitty — CHANGES_REQUESTED** [review]
Looks good overall! That said, I think it would be helpful to refine the types and avoid `squeeze` where possible. See individual comments for details.

File: pyciemss/ouu/ouu.py (hunk 1 above)
Comment (anirban-chaudhuri) [reply pyciemss/ouu/ouu.py:74]: We could but I wonder what would be the downstream effects of changing the name here. I am guessing it should be doable...

File: pyciemss/ouu/ouu.py (hunk 2 above)
Comment (anirban-chaudhuri) [reply pyciemss/ouu/ouu.py:120]: Yes, the risk_estimate is an array now and scipy can handle the numpy array to create multiple constraints.

File: pyciemss/interfaces.py [hunk 3]
```diff
@@ -529,12 +529,15 @@ def wrapped_model():
...
+            k: (v.squeeze() if len(v.shape) > 2 else v) for k, v in samples.items()
+        }
 
         risk_results = {}
         for k, vals in samples.items():
             if "_state" in k:
-                # qoi is assumed to be the last day of simulation
-                qoi_sample = vals.detach().numpy()[:, -1]
+                # Note: qoi is assumed to be the last day of simulation
+                qoi_sample = np.atleast_2d(np.squeeze(vals.detach().numpy()))[:, -1]
```
[the remark this answers is not in the document]
Comment (anirban-chaudhuri) [reply pyciemss/interfaces.py:540]: When I am using detach, it is introducing the extra dimension for certain cases. I am trying to track down the issue with the formatting of interventions, but haven't had enough time. Once that it fixed we should be able to get rid of the sqeeze. However, I think it is a separate issue unrelated to introducing multiple constraints for `optimize` and is better handled in a separate PR.

File: pyciemss/interfaces.py [hunk 4]
```diff
@@ -863,21 +868,35 @@ def optimize(
                     - Optimization results as scipy object.
     """
     check_solver(solver_method, solver_options)
+    if not isinstance(risk_bound, list):
```
[the remark this answers is not in the document]
Comment (anirban-chaudhuri) [reply pyciemss/interfaces.py:871]: That is true. I put it in so that we don't get an error if it is not a list. I will change the type hint.

File: pyciemss/interfaces.py [hunk 5]
```diff
@@ -803,9 +806,9 @@ def optimize(
             - The end time of the sampled simulation.
         logging_step_size: float
             - The step size to use for logging the trajectory.
-        qoi: Callable
+        qoi: List[Callable[[Any], np.ndarray]]
```
[the remark this answers is not in the document]
Comment (anirban-chaudhuri) [reply pyciemss/interfaces.py:809]: I added that it could be a list to the documentation. I also updated the type hint to be a Union[]

File: pyciemss/integration_utils/intervention_builder.py [hunk 6]
```diff
@@ -49,6 +50,8 @@ def start_time_objective(
     param_value: List[Intervention],
 ) -> Callable[[torch.Tensor], Dict[float, Dict[str, Intervention]]]:
     param_size = len(param_name)
+    # Note: code below will only work for tensors and not callable functions
+    param_value = [torch.atleast_1d(y) for y in param_value]
```
[the remark this answers is not in the document]
Comment (anirban-chaudhuri) [reply pyciemss/integration_utils/intervention_builder.py:54]: I changed this one. However, the rest can still handle type `Intervention`.

File: pyciemss/ouu/ouu.py [hunk 7]
```diff
@@ -162,6 +164,9 @@ def wrapped_model():
                     num_samples=self.num_samples,
                     parallel=True,
                 )()
+        samples = {
+            k: (v.squeeze() if len(v.shape) > 2 else v) for k, v in samples.items()
```
[the remark this answers is not in the document]
Comment (anirban-chaudhuri) [reply pyciemss/ouu/ouu.py:168]: This particular line is from https://github.com/ciemss/pyciemss/blob/f5c5bec731441377b845a3b702f01927113bf462/pyciemss/integration_utils/result_processing.py#L47
Should we not have it there as well?

File: docs/source/interfaces.ipynb
Comment (anirban-chaudhuri) [reply docs/source/interfaces.ipynb:1]: Let's discuss this `squeeze` issue and see what we want to do.

File: pyciemss/interfaces.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (anirban-chaudhuri) [reply pyciemss/interfaces.py:540]: Done! Thanks!

File: pyciemss/ouu/ouu.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (anirban-chaudhuri) [reply pyciemss/ouu/ouu.py]: We need to fix this in a separate PR. I have removed the squeeze from OUU codes.

File: pyciemss/ouu/qoi.py [hunk 8]
```diff
@@ -25,5 +25,5 @@ def obs_max_qoi(samples: Dict[str, torch.Tensor], contexts: List) -> np.ndarray:
     samples[VARIABLE] is expected to have dimension (nreplicates, ntimepoints)
     """
     dataQoI = samples[contexts[0]].detach().numpy()
-
+    dataQoI = np.atleast_2d(np.squeeze(dataQoI))
```
[the remark this answers is not in the document]
Comment (anirban-chaudhuri) [reply pyciemss/ouu/qoi.py]: Removed squeeze

File: docs/source/interfaces.ipynb
Comment (anirban-chaudhuri) [reply docs/source/interfaces.ipynb:1]: Removed squeeze

File: docs/source/optimize_interface.ipynb
Comment (anirban-chaudhuri) [reply docs/source/optimize_interface.ipynb:1]: Removed squeeze

**Comment by anirban-chaudhuri** [comment]
> Looks good overall! That said, I think it would be helpful to refine the types and avoid `squeeze` where possible. See individual comments for details.

All the `squeeze` have now been removed from `optimize`, but there is still one in results processing which should be a separate PR.

**Review by SamWitty — APPROVED** [review]
Looks great!

**Comment by liunelson** [comment]
@anirban-chaudhuri 
Could you confirm that I'm using your new interface correctly?

Suppose I want the following "success criteria" or "constraints":
* Ensure `S` is below `1` at all timepoints (risk level 95%)
* Ensure `S` is above `1` at the last timepoint (risk level 95%)
image

```python
observed_params = [["S_state"], ["S_state"]]
risk_bound = [1.0, -2.0]
alpha = [0.95, 0.95]
qoi = [
    lambda y: obs_max_qoi(y, observed_params[0]),
    lambda y: -obs_nday_average_qoi(y, observed_params[1])
]
```

Is the minus sign on the risk bound and `qoi` function the right way to specify an "above threshold" constraint?

**Comment by liunelson** [comment]
Also, how is `0.4` in `scaling_factor_obj[1]` chosen in cell 25?
https://github.com/liunelson/pyciemss/blob/nl-test/docs/source/optimize_interface.ipynb

I'd have expected it to be just `1.0 / (bounds_interventions[1][1] - bounds_interventions[0][1])`
